core/src/main/java/org/jruby/util/Dir.java

```diff
@@ -776,50 +810,39 @@ private static int addToResultIfExists(Ruby runtime, String cwd, byte[] bytes, i
         FileResource file = JRubyFile.createResource(runtime, cwd, fileName);
 
         if (file.exists()) {
-            byte[] newBytes;
-            
-            // get the real filename (case-sensitive)
-            if(file.isFile() && cwd != null) {
-                String path = file.isSymLink() ? file.absolutePath() : file.canonicalPath();
-
-                // compensate for missing slash
-                if (fileName.endsWith("/")) {
-                    path += "/";
-                }
-
-                if(fileName.contains("./")) {
-                    newBytes = ("./" + path.substring((path.length() - end + 2))).getBytes();
-                } else {
-                    int tempBegin = path.length() - fileName.length();
-                    newBytes = path.substring(tempBegin).getBytes();
-                }
-
-                end = newBytes.length;
-            } else {
-                newBytes = bytes;
-            }
-
-            return func.call(newBytes, begin, end - begin, enc, arg);
+            return func.call(bytes, begin, end - begin, enc, arg);
         }
 
         return 0;
     }
 
     private static int glob_helper(Ruby runtime, String cwd, ByteList path, int sub, int flags, GlobFunc<GlobArgs> func, GlobArgs arg) {
+        int begin = path.getBegin();
+        final int end = begin + path.length();
+        final Encoding enc = path.getEncoding();
+        final byte[] bytes = path.getUnsafeBytes();
+        final byte[] scheme = extractScheme(bytes, begin, end);
+        if (scheme != null) begin += scheme.length;
+        return glob_helper(runtime, cwd, scheme, bytes, begin, end, enc, sub, flags, func, arg);
+    }
+
+    private static int glob_helper(Ruby runtime, String cwd, byte[] scheme, ByteList path, int sub, int flags, GlobFunc<GlobArgs> func, GlobArgs arg) {
         final int begin = path.getBegin();
-        final int end = begin + path.getRealSize();
+        final int end = begin + path.length();
         final Encoding enc = path.getEncoding();
-        return glob_helper(runtime, cwd, path.getUnsafeBytes(), begin, end, enc, sub, flags, func, arg);
+        return glob_helper(runtime, cwd, scheme, path.getUnsafeBytes(), begin, end, enc, sub, flags, func, arg);
     }
 
-    private static int glob_helper(Ruby runtime, String cwd,
+    private static int glob_helper(Ruby runtime, String cwd, byte[] scheme,
         byte[] path, int begin, int end, Encoding enc, int sub,
         int flags, GlobFunc<GlobArgs> func, GlobArgs arg) {
         int status = 0;
 
         int ptr = sub != -1 ? sub : begin;
 
-        if ( ! has_magic(path, ptr, end, flags) ) {
+        final GlobMagic nonMagic = CASEFOLD_FILESYSTEM ? GlobMagic.PLAIN : GlobMagic.ALPHA;
+
+        if ( has_magic(path, ptr, end, flags).compareTo(nonMagic) <= 0 ) {
```

Comment:
This change seems to mean we'll want to ensure we preserve enum order. Hopefully we're not likely to add any more? Probably should have some documentation to indicate that the order of that enum is important, or else add a method to the enum that more clearly spells out how the "priority" of the magic is determined.

Reply:
MRI has more, but it handles braces and brackets differently

Reply:
I've added a comment stating order is important